sled-agent/src/illumos/running_zone.rs

```diff
@@ -345,10 +399,17 @@ impl InstalledZone {
         let zone_image_path =
             PathBuf::from(&format!("/opt/oxide/{}.tar.gz", service_name));
 
+        let phys_iter = if let Some(ref vnic) = physical_vnic {
+            vec![vnic.name().to_string()].into_iter()
+        } else {
+            vec![].into_iter()
```

Comment:
`std::iter::empty()` might be nice, but not important.

Reply:
Annoyingly, this won't work because it would mean `phys_iter` would need to have different types depending on which branch it took (`vec::IntoIter` vs `iter::Empty`). Making `phys_iter` a `Box<dyn Iterator<Item=_>>` would work around that, but it would force a heap allocation that doesn't happen with the code as written, since an empty vec is guaranteed not to allocate.

Reply:
Ugh, right. I forget that `empty` returns a separate type here. Thanks John.

Reply:
This exact thing bites me at least a few times a month!

Looking back at the code, we could (ab)use the fact that `Option` implements iterator to get rid of the length 1 vec in the true case. Looks weird, but this is equivalent with no temporary vec:

```rust
let phys_iter = if let Some(ref vnic) = physical_vnic {
    Some(vnic.name().to_string()).into_iter()
} else {
    None.into_iter()
};
```

Or maybe at that point it could even been inlined below?

```rust
        let net_device_names: Vec<String> = opte_ports
￼            .iter()
￼            .map(|port| port.vnic().name().to_string())
￼            .chain(std::iter::once(control_vnic.name().to_string()))
￼            .chain(physical_vnic.map(|vnic| vnic.name().to_string()))
￼            .collect();
```

Reply:
Updated to the inlined version!